lib/gateway_traversal.go

```diff
@@ -0,0 +1,284 @@
+package lib
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	bsfetcher "github.com/ipfs/boxo/fetcher/impl/blockservice"
+	"github.com/ipfs/boxo/gateway"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-unixfsnode"
+	"github.com/ipfs/go-unixfsnode/data"
+	"github.com/ipld/go-car"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/ipld/go-ipld-prime/traversal"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	selectorparse "github.com/ipld/go-ipld-prime/traversal/selector/parse"
+)
+
+type getBlock func(ctx context.Context, cid cid.Cid) (blocks.Block, error)
+
+func carToLinearBlockGetter(ctx context.Context, reader io.Reader, metrics *GraphGatewayMetrics) (getBlock, error) {
+	cr, err := car.NewCarReaderWithOptions(reader, car.WithErrorOnEmptyRoots(false))
+	if err != nil {
+		return nil, err
+	}
+
+	cbCtx, cncl := context.WithCancel(ctx)
+
+	type blockRead struct {
+		block blocks.Block
+		err   error
+	}
+
+	blkCh := make(chan blockRead, 1)
+	go func() {
+		defer cncl()
+		defer close(blkCh)
+		for {
+			blk, rdErr := cr.Next()
+			select {
+			case blkCh <- blockRead{blk, rdErr}:
+				if rdErr != nil {
+					cncl()
+				}
+			case <-cbCtx.Done():
+				return
+			}
+		}
+	}()
+
+	isFirstBlock := true
+	mx := sync.Mutex{}
+
+	return func(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+		mx.Lock()
+		defer mx.Unlock()
+
+		// initially set a higher timeout here so that if there's an initial timeout error we get it from the car reader.
+		var t *time.Timer
+		if isFirstBlock {
+			t = time.NewTimer(GetBlockTimeout * 2)
+		} else {
+			t = time.NewTimer(GetBlockTimeout)
+		}
+		var blkRead blockRead
+		var ok bool
+		select {
+		case blkRead, ok = <-blkCh:
+			if !t.Stop() {
+				<-t.C
+			}
+			t.Reset(GetBlockTimeout)
+		case <-t.C:
+			return nil, gateway.ErrGatewayTimeout
+		}
+		if !ok || blkRead.err != nil {
+			if !ok || errors.Is(blkRead.err, io.EOF) {
+				return nil, io.ErrUnexpectedEOF
+			}
+			return nil, GatewayError(blkRead.err)
+		}
+		if blkRead.block != nil {
+			metrics.carBlocksFetchedMetric.Inc()
+			if !blkRead.block.Cid().Equals(c) {
+				return nil, fmt.Errorf("unexpected block received: expected %s, got %s", c, blkRead.block.Cid())
```

Comment:
Opening  `http://en.wikipedia-on-ipfs.org.ipns.localhost:8081/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp` produces request for `/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity` and that errors with  `failed to resolve /ipns/en.wikipedia-on-ipfs.org/I/Cour_de_Marbre_du_Ch\u00e2teau_de_Versailles_October_5%2C_2011.jpg.webp: unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py`


This happens three times with different L1s:

```
2023-08-03T22:15:25.570+0200	DEBUG	namesys	namesys/base.go:75	resolved /ipns/en.wikipedia-on-ipfs.org to /ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze
2023-08-03T22:15:25.570+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:105	doing fetch	{"from": "145.239.19.141", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "mime": "application/vnd.ipld.car", "requestId": "97e6b448-4cda-424d-b1ef-a89171199997"}
2023-08-03T22:15:25.608+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:156	fetch result	{"from": "145.239.19.141", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "status": 200, "size": 32356, "duration": 0.038247819, "attempt": 0, "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py", "proto": "HTTP/2.0"}
2023-08-03T22:15:25.608+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/pool.go:549	fetch attempt failed	{"from": "145.239.19.141", "attempt": 0, "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py"}
2023-08-03T22:15:25.609+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:105	doing fetch	{"from": "145.239.19.144", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "mime": "application/vnd.ipld.car", "requestId": "59f59d5b-36b1-4116-8da1-78583c5df23e"}
2023-08-03T22:15:25.685+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:156	fetch result	{"from": "145.239.19.144", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "status": 200, "size": 32356, "duration": 0.076737074, "attempt": 1, "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py", "proto": "HTTP/2.0"}
2023-08-03T22:15:25.685+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/pool.go:549	fetch attempt failed	{"from": "145.239.19.144", "attempt": 1, "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py"}
2023-08-03T22:15:25.686+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:105	doing fetch	{"from": "51.77.48.117", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "mime": "application/vnd.ipld.car", "requestId": "adbcaad5-75b0-4b66-b132-69c638514a1e"}
2023-08-03T22:15:25.799+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/fetcher.go:156	fetch result	{"from": "51.77.48.117", "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "status": 200, "size": 32356, "duration": 0.113721215, "attempt": 2, "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py", "proto": "HTTP/2.0"}
2023-08-03T22:15:25.799+0200	DEBUG	caboose	caboose@v0.0.3-no-response-drain-2/pool.go:549	fetch attempt failed	{"from": "51.77.48.117", "attempt": 2, "of": "/ipfs/bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze/I/Cour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%252C_2011.jpg.webp?format=car&dag-scope=entity", "error": "unexpected block received: expected bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u, got bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py"}
```

and user is unable to get requested data:

>  ![2023-08-03_22-09](https://github.com/ipfs/bifrost-gateway/assets/157609/0d4b55a3-cdae-40c2-aef5-7a48e06a9d91)

@aschmahmann  is the lack of block-by-block fallback intentional (hard fail to force Saturn to behave)?

Reply:
i think this is being used currently to make sure the incremental validation is working correctly.

it does seem not unreasonable to have a flag to allow fallback to block-by-block in case the backend car delivery fails for specific requests?

we should understand where/why we're erroring here though.

I wouldn't be shocked if you told me the error was that we were getting the direct object asked for without the directory blocks that we needed to verify it, for instance.

Reply:
yeah, the lack of block-by-block requests was intentional as it should be safe to assume that well behaving peers are doing the protocol correctly. We could re-add the fallbacks at the cost of more code and complexity if we wanted to, but either way we've got some mismatch between what the client expects and what the server is returning and one of them is wrong 😄. We should figure out which and then fix it.

Reply:
The production reality is that we will never be able to remove this problem fully: anyone can run "misconfigured" L1 and there is always a possibility of invalid response.  

Production at ipfs.io will require us to either:
-  have fallback to block-by-block enabled
   - if it is controlled via [environment variable](https://github.com/ipfs/bifrost-gateway/blob/main/docs/environment-variables.md), we can run without it in staging to catch issues
- and/or a way to let caboose know we don't want to talk to a L1 or at least signal back it returned invalid CAR
  - if I could retry with different L1s i might get valid CAR data from different L1 without falling back to block-by-block. 
  - perhaps this is less work for now?

Reply:
I ran bifrost-gateway locally against both Saturn (core nodes) and kubo and going to that URL worked fine. Note: the expected traversal is:

```
bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze -> wiki root
bafybeibuk3ivdskzskdzt2ru66vqycfeyn4fizkpcimlc3h2f672fuvkhi -> I dir
bafybeiakqoj432t4evvkaw72uv6yireyfi2dmkbow4fdpc7zissd3sif7u -> A9 internal HAMT DIR
bafybeig2l6gelwzahczw3gaxg6tinqqcanceoucoijg2wdh4nmmvvhc7ie -> 9D internal HAMT DIR
bafkreiexrc5ydhcftjravsjnjmvojn3aexkbevq3ai5onf7vivwtlmkufu -> 1CCour_de_Marbre_du_Ch%C3%A2teau_de_Versailles_October_5%2C_2011.jpg.webp
```

`bafybeih3o36bjvtxfv2hr7lkalxh4qh4d7rwscn6kfp3huwghszf3d35py`
Is in the first HAMT layer under I, but instead of A9 it's BA. Not totally sure why that'd be returned though... maybe a misinterpretation of the escaping (i.e. it ends up unescaped and the murmur3 hash of the unescaped version starts looking in bucket BA rather than A9).